### ISSUE TITLE
Panic: index out of range when extracting scalar value

### ISSUE THREAD

**Issue by Byter09** [influxdata/flux#4692]
Hello everyone!

I've been using Flux for a bit now and I hope I'm reporting this in the correct repository.

Basically, I managed to cause a panic simply by calling a function that returns a row inside of a map:

```
callsValueBelowPrice = (price) =>
  intrinsic_value 
    |> filter(fn: (r) => r.type == "call")
    |> filter(fn: (r) => float(v: r.price) == price)
    |> filter(fn: (r) => float(v: r.strike) < price)
    |> keep(columns: ["cash_value"])
    |> sum(column: "cash_value")
    |> findRecord(fn: (key) => true, idx: 0)

strikes
  |> map(fn: (r) => ({r with calls_value: callsValueBelowPrice(price: r.price).cash_value}))
  ```

I'm unsure of what I can remove above to reduce the example, but I hope that's okay.

Here's something that _does not_ do what I want, but it doesn't panic:

```
cv = callsValueBelowPrice(price: 100).cash_value

strikes
  |> map(fn: (r) => ({r with calls_value: cv}))
```

As soon as I even try to put this call in the map function in any way, I get a panic.

Even this panics:
```
callsValueBelowPrice = (price) => {
  sum = intrinsic_value 
    |> filter(fn: (r) => r.type == "call")
    |> filter(fn: (r) => float(v: r.price) == price)
    |> filter(fn: (r) => float(v: r.strike) < price)
    |> keep(columns: ["cash_value"])
    |> sum(column: "cash_value")
    |> findRecord(fn: (key) => true, idx: 0)

  return sum.cash_value
}

strikes
  |> map(fn: (r) => ({r with calls_value: callsValueBelowPrice(price: r.price)}))
```

Here's the error I get in the console:
```
ts=2022-04-21T20:09:45.371737Z lvl=info msg="Dispatcher panic" log_id=0_zSXqpl000 service=storage-reads component=dispatcher error="panic: runtime error: index out of range [2] with length 2" stacktrace="goroutine 12548271 [running]:
runtime/debug.Stack()
    /go/src/runtime/debug/stack.go:24 +0x65
github.com/influxdata/flux/execute.(*poolDispatcher).recover(0xc018a923f0)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/recover.go:53 +0x205
panic({0x3c103c0, 0xc02e13f7e8})
    /go/src/runtime/panic.go:1038 +0x215
github.com/influxdata/flux/stdlib/universe.createMergeJoinTransformation({0x51, 0x76, 0xf, 0x64, 0xe4, 0x7b, 0x54, 0xd4, 0xbb, 0x7b, ...}, ...)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/stdlib/universe/join.go:220 +0x49d
github.com/influxdata/flux/execute.(*createExecutionNodeVisitor).Visit(0xc0298c81e0, {0x3e18868, 0xc012a5eab0})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/executor.go:251 +0x723
github.com/influxdata/flux/plan.walk({0x3e18868, 0xc012a5eab0}, 0x3d49b28, 0xc0298c8200, 0x41758b)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/plan/walk.go:88 +0xf2
github.com/influxdata/flux/plan.walk({0x3e18868, 0xc012a5ea20}, 0x3d49b28, 0xc0298c8200, 0x41758b)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/plan/walk.go:82 +0x133
github.com/influxdata/flux/plan.walk({0x3e18868, 0xc012a5e990}, 0x3d49b28, 0xc0298c8200, 0x41758b)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/plan/walk.go:82 +0x133
github.com/influxdata/flux/plan.walk({0x3e18868, 0xc012a5e900}, 0x3d49b28, 0xc0298c8200, 0x41758b)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/plan/walk.go:82 +0x133
github.com/influxdata/flux/plan.walk({0x3e18868, 0xc012a5e870}, 0x3d49b28, 0xc0298c8200, 0x41758b)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/plan/walk.go:82 +0x133
github.com/influxdata/flux/plan.walk({0x3e18868, 0xc012a5e7e0}, 0x3d49b28, 0xc0298c8200, 0x41758b)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/plan/walk.go:82 +0x133
github.com/influxdata/flux/plan.walk({0x3e18868, 0xc012a5e750}, 0x3d49b28, 0xc0298c8200, 0x417619)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/plan/walk.go:82 +0x133
github.com/influxdata/flux/plan.walk({0x3e18868, 0xc012a5e6c0}, 0x3d49b28, 0xc0298c8200, 0x0)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/plan/walk.go:82 +0x133
github.com/influxdata/flux/plan.walk({0x3e18868, 0xc012a5e630}, 0x3d49b28, 0xc0298c8200, 0x10000c04360770f)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/plan/walk.go:82 +0x133
github.com/influxdata/flux/plan.(*Spec).BottomUpWalk(0xc020485440, 0xc043607590)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/plan/walk.go:60 +0x2a8
github.com/influxdata/flux/execute.(*executor).createExecutionState(0xc0298c8360, {0x3dd2520, 0xc043607590}, 0xc020485440, {0x3dd3390, 0xc012b809b0})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/executor.go:99 +0x225
github.com/influxdata/flux/execute.(*executor).Execute(0x3986320, {0x3dd2520, 0xc043607590}, 0xf, {0x3dd3390, 0xc012b809b0})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/executor.go:73 +0x3d
github.com/influxdata/flux/lang.(*Program).Start(0xc043607560, {0x3dd2520, 0xc04355b0e0}, {0x3dd3390, 0xc03b17e7d0})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/lang/compiler.go:310 +0x7c5
github.com/influxdata/flux/stdlib/universe.tableFind({0x3dd2520, 0xc04355b0e0}, 0xc00c49a980, 0xc04355b140)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/stdlib/universe/table_fns.go:101 +0x2d6
github.com/influxdata/flux/stdlib/universe.findRecordCall({0x3dd2520, 0xc04355b0e0}, {0x3e24598, 0xc012b80820})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/stdlib/universe/table_fns.go:419 +0x3fd
github.com/influxdata/flux/values.(*function).Call(0x0, {0x3dd2520, 0xc04355b0e0}, {0x3e24598, 0xc012b80820})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/values/function.go:119 +0x3e
github.com/influxdata/flux/interpreter.(*Interpreter).doCall(0xc01105aac0, {0x3dd2478, 0xc0110a83c0}, 0xc010efd4d0, {0x3e11248, 0xc04355ac00})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:724 +0x4e3
github.com/influxdata/flux/interpreter.(*Interpreter).doExpression(0xc01105aac0, {0x3dd2478, 0xc0110a83c0}, {0x3de3d78, 0xc010efd4d0}, {0x3e11248, 0xc04355ac00})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:368 +0x129f
github.com/influxdata/flux/interpreter.(*Interpreter).doStatement(0xc04355aae0, {0x3dd2478, 0xc0110a83c0}, {0x3dde838, 0xc03b17f540}, {0x3e11248, 0xc04355ac00})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:210 +0x18f
github.com/influxdata/flux/interpreter.function.doCall({0xc010f03c80, {0x3e11248, 0xc01104a0c0}, 0xc01105aac0}, {0x3dd2478, 0xc0110a83c0}, {0x3e25038, 0xc0521d0f60})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:938 +0x230
github.com/influxdata/flux/interpreter.function.Call({0xc010f03c80, {0x3e11248, 0xc01104a0c0}, 0xc01105aac0}, {0x3dd2478, 0xc0110a83c0}, {0x3e24598, 0xc012b80000})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/interpreter/interpreter.go:886 +0x165
github.com/influxdata/flux/compiler.(*callEvaluator).Eval(0xc0298b20c0, {0x3dd2478, 0xc0110a83c0}, {0x3e11f60, 0xc024f24d30})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/compiler/runtime.go:651 +0x3bc
github.com/influxdata/flux/compiler.(*memberEvaluator).Eval(0xc0298ad3b0, {0x3dd2478, 0xc0110a83c0}, {0x3e11f60, 0xc024f24d30})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/compiler/runtime.go:560 +0x7f
github.com/influxdata/flux/compiler.eval(...)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/compiler/runtime.go:88
github.com/influxdata/flux/compiler.(*objEvaluator).Eval.func1(0xc0521d0f00)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/compiler/runtime.go:223 +0x30d
github.com/influxdata/flux/values.BuildObjectWithSize(0x0, 0xc0298c92b0)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/values/object.go:113 +0x172
github.com/influxdata/flux/values.BuildObject(...)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/values/object.go:100
github.com/influxdata/flux/compiler.(*objEvaluator).Eval(0x58c3b50, {0x3dd2478, 0xc0110a83c0}, {0x3e11f60, 0xc024f24d30})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/compiler/runtime.go:205 +0x7c
github.com/influxdata/flux/compiler.eval(...)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/compiler/runtime.go:88
github.com/influxdata/flux/compiler.(*blockEvaluator).Eval(0xc012b75f90, {0x3dd2478, 0xc0110a83c0}, {0x3e11f60, 0xc024f24d30})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/compiler/runtime.go:112 +0x98
github.com/influxdata/flux/compiler.eval(...)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/compiler/runtime.go:88
github.com/influxdata/flux/compiler.compiledFn.Eval({{0x3da5cc0, 0xc0298b2180}, {0x3e11f60, 0xc01108f720}}, {0x3dd2478, 0xc0110a83c0}, {0x3e24598, 0xc012b75f90})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/compiler/runtime.go:46 +0x19c
github.com/influxdata/flux/execute.(*rowFn).eval(0xc020484900, {0x3dd2478, 0xc0110a83c0}, 0xc0000a2101, {0x3e110f8, 0xc02987b800}, 0x434f92)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/row_fn.go:251 +0x11d
github.com/influxdata/flux/execute.(*RowMapPreparedFn).Eval(0x4663c0, {0x3dd2478, 0xc0110a83c0}, 0xc00fb33040, {0x3e110f8, 0xc02987b800})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/row_fn.go:341 +0x32
github.com/influxdata/flux/stdlib/universe.(*mapTransformation).Process.func1({0x3e110f8, 0xc02987b800})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/stdlib/universe/map.go:160 +0xd4
github.com/influxdata/flux/execute.(*consecutiveTransportTable).Do.func1({0x3e110f8, 0xc02987b800})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/transport.go:538 +0x542
github.com/influxdata/flux/execute/table.(*BufferedTable).Do(0x3bce5a0, 0xc0521d0ea0)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/table/buffered.go:69 +0x19b
github.com/influxdata/flux/execute.(*consecutiveTransportTable).Do(0xc035c2fd28, 0xc04355a900)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/transport.go:520 +0x87
github.com/influxdata/flux/stdlib/universe.(*mapTransformation).Process(0xc03efc65a0, {0x8c, 0x4, 0x88, 0x37, 0x9c, 0x7e, 0x59, 0x12, 0x86, ...}, ...)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/stdlib/universe/map.go:157 +0x10a
github.com/influxdata/flux/execute.(*transformationTransportAdapter).ProcessMessage(0xc0110ac1e0, {0x3dd3208, 0xc0298a1e20})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/transport.go:618 +0x19b
github.com/influxdata/flux/execute.(*consecutiveTransport).processMessage(0xc012ed3760, {0x3dd2478, 0xc0110a83c0}, {0x3dd3208, 0xc0298a1e20})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/transport.go:263 +0x104
github.com/influxdata/flux/execute.(*consecutiveTransport).processMessages(0xc012ed3760, {0x3dd2478, 0xc0110a83c0}, 0xa)
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/transport.go:218 +0xe6
github.com/influxdata/flux/execute.(*poolDispatcher).doWork(0xc018a923f0, {0x3dd2478, 0xc0110a83c0})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/dispatcher.go:157 +0x19f
github.com/influxdata/flux/execute.(*poolDispatcher).run(0xc018a923f0, {0x3dd2478, 0xc0110a83c0})
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/dispatcher.go:135 +0x4f
github.com/influxdata/flux/execute.(*poolDispatcher).Start.func1()
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/dispatcher.go:75 +0x86
created by github.com/influxdata/flux/execute.(*poolDispatcher).Start
    /go/pkg/mod/github.com/influxdata/flux@v0.161.0/execute/dispatcher.go:71 +0x4d
"
```

I tried looking for this error for hours now but couldn't find anything, so I hope this isn't a duplicate.

In the meantime, is there some kind of workaround I can apply? The query I'm writing is rather complicated and I can't think of another way to solve this, as `map()` is pretty essential here.

**Comment by athornton** [comment]
Posted on Slack, but it might be relevant here.  Or might not:

Looks kinda similar to https://github.com/influxdata/flux/pull/3255/ .

That was fixed by adding a test, join_right_side_more_cols_test.flux to join when it was in experimental, but I don't see any such test name in https://github.com/influxdata/flux/tree/master/stdlib/universe .

So I wonder if somehow that test (and functionality) didn't make it forward out of experimental.

**Comment by Byter09** [comment]
I was being told, as there is no `join` in the above query, that I should post the whole query.

Here's the whole thing, comments and all. I _really_ have no clue how to minimize this.

```
import "math"

data = from(bucket: "finance")
  |> range(start: v.timeRangeStart, stop: v.timeRangeStop)
  |> filter(fn: (r) => r["_measurement"] == "options")
  |> filter(fn: (r) => r["isin"] == "US36467W1099")
  //|> filter(fn: (r) => r["expiration"] == "${expiration}")
  |> filter(fn: (r) => r["expiration"] == "1650585600000")
  //|> filter(fn: (r) => if "${strike}" == "all" then true else r["strike"] =~ /^${strike:regex}$/)
  |> last()

// from https://maximum-pain.com/blog/archive/How-to-Calculate-Max-Pain
oi = data
  |>filter(fn: (r) => r["_field"] == "oi")
  |> rename(columns: {_value: "oi"})
  |> keep(columns: ["oi", "isin", "strike", "type"])

strikes = data
  |> keep(columns: ["strike", "isin"])
  |> unique(column: "strike")
  |> rename(columns: {strike: "price"})


intrinsic_value = join(tables: {oi: oi, strikes: strikes}, on: ["isin"], method: "inner")
  |> drop(columns: ["isin"])
  |> map(fn: (r) => ({
    r with
      intrinsic: math.mMax(x: if r.type == "call" then float(v: r.price) - float(v: r.strike) else float(v: r.strike) - float(v: r.price), y: 0.0),
  }))
  |> map(fn: (r) => ({ r with cash_value: r.oi * 100.0 * r.intrinsic }))

// get all calls that have a lower strike but that "price"
callsValueBelowPrice = (price) => {
  sum = intrinsic_value 
    |> filter(fn: (r) => r.type == "call")
    |> filter(fn: (r) => float(v: r.price) == price)
    |> filter(fn: (r) => float(v: r.strike) < price)
    |> keep(columns: ["cash_value"])
    |> sum(column: "cash_value")
    |> findRecord(fn: (key) => true, idx: 0)

  return sum.cash_value
}

strikes
  |> map(fn: (r) => ({r with calls_value: callsValueBelowPrice(price: r.price)}))
```

**Comment by Byter09** [comment]
And, finally, so you guys don't have to fiddle in the dark, here's what "data" returns. That way you can test all the following functions yourself.

[2022-04-21_23 32_influxdb_data.csv](https://github.com/influxdata/flux/files/8535222/2022-04-21_23.32_influxdb_data.csv)

**Comment by mhall119** [comment]
Could possibly be related to https://community.influxdata.com/t/nested-query-runs-out-of-memory/24735

**Comment by onelson** [comment]
:pray: thanks, @Byter09. I'm taking a look.

**Comment by onelson** [comment]
No fixes yet, but I was able to reproduce the panic in flux@master with the provided script and data. Mostly I'm taking notes here to help me pick this back up on Monday.

Walking through in a debugger I see that eventually we hit a point where `s.TableNames` here has a length of 4, but we have a requirement that `parents` must be no larger than 2. We see this enforced just above the trigger of the panic:
https://github.com/influxdata/flux/blob/ed35a7b1078902a92da18dbb8c5be2091b6f2bcd/stdlib/universe/join.go#L213-L221

I'll need to study the flux program a bit, actually reading it :rofl:, so as to get a sense of how we get here and what the correction should be.

**Comment by onelson** [comment]
Managed to simplify the repro case as:

```
import "array"


xs = array.from(rows: [{id: 1, x: 1}, {id: 2, x: 2}])
ys = array.from(rows: [{id: 1, y: 1}, {id: 2, y: 2}])
zs = join(tables: {xs: xs, ys: ys}, on: ["id"])

getById = (id) => {
  r = zs |> filter(fn: (r) => r.id == id) |> findRecord(fn: (key) => true, idx: 0)
  return { x: r.x, y: r.y }
}

array.from(rows: [{ _value: 1}, { _value: 2}])
  |> map(fn: (r) => getById(id: r._value))
```

The gist is when `findRecord` runs repeatedly on the same stream of tables, and those tables are built using `join`, the spec for the join node gets mutated and starts including repeats of the incoming table names. In this case, my 2nd iteration shows the join spec as listing:
`xs`, `xs`, `ys`, `ys`. This leads to the panic.

In an unrelated discussion with the team earlier, it came up that _"repeatedly running `findRecord` on the same stream can have unknown consequences."_ Seems like this is one of them.

---

A possible workaround for this specific case could be to make `intrinsic_value` a function that returns a fresh copy of the joined stream each time the `map()` function is invoked, but this will be _very slow_, I'm sure. I ran the script with this change and it went for several minutes given the csv data attached to the issue above. Not great :disappointed:.

I'm going to work towards ensuring we convert the panic to an error if it is truly unavoidable due to the semantics of `findRecord`. I'll also attempt to see if we can make this work in a safe way -- TBD.

**Comment by Byter09** [comment]
Is there a speedier alternative that I can use? All I want to do is:
- get variables using a query (in my example strikes)
- use variable in secondary query
- aggregate data in that query (sum in my case)
- store result in field of row that gave me the original variable

I can't think of any other way than the approach I used, which evidently panics.
And even if you convert that into an error, it wouldn't really help me at all :(

**Comment by onelson** [comment]
> And even if you convert that into an error, it wouldn't really help me at all :(

Sure. The _error vs panic as-is_ concern is mostly just about how this shows up in logs, hopefully steering us to the cause more directly. It took me some legwork to figure out what the story was initially, so I was aiming to tighten that up.

> Is there a speedier alternative that I can use?

I'm not sure. A pattern that I've seen used, though not exactly for this purpose, is to pre-process the data and write it into a bucket as a part of a Task then read from that bucket in a subsequent query.

On the whole, I don't know of great patterns for this but folks in slack may have ideas.

---

I've posted a PR that should resolve the panic, but I'm not 100% sure the fix is entirely healthy -- will need to confer with the team. With that patch applied, I re-ran your script against the csv data and the runtime was roughly 8m30s -- I think this was similar to my initial workaround of wrapping the joined data in a function so each `findRecord` call received a "fresh" version of the stream.

I think something to keep in mind is that `findRecord` is essentially taking the stream you feed into it and building a sub-program from it, then executing that program and waiting for the output. I think there's going to be an non-insignificant cost for each time this happens (ie, for each item you're mapping over).
Something like the proposed [`array.to`](https://github.com/influxdata/flux/issues/3424) might be helpful in this situation. The idea there was to add a way to _force a stream to materialize_, making it possible to use it for subsequent lookups (ie, by building a `dict`).
Unfortunately, it's still just a proposal so it's not going to be any help today :sob:

**Comment by Byter09** [comment]
I don't have enough knowledge of the internals of flux to come even close to understanding this 😅

I can however say that my final query (once I can continue working on it), will do this not only once and for one field per row, it'll do it twice for all rows because I also need to sum up all option puts above the strike. And those are way, waaay more than the calls. So I feel like this is going to be one lengthy query, in terms of execution time.

I will however think about using the tasks approach. Given that I don't need to calculate this continuously (every hour is okay-ish, even though I would prefer per-minute), that should give me at least a result. Testing is going to be pain though heh.

I don't quite know what the bottleneck is (as I don't understand it fully), but I trust that you'll find a way in future versions.

My last concern is ram usage. I run my queries on a server with only 16GB of ram, InfluxDB is already taking up quite a bit of that. I'm a systems programmer using Rust, so memory usage is kind of always on my mind. The amount of "cloning" I've read so far scares me a little. Will my query blow up and force the OOM to kick in?


Nevertheless thank you so much for taking the time to look at this and for finding a "hotfix" at least. 🥳

**Comment by onelson** [comment]
> The amount of "cloning" I've read so far scares me a little. Will my query blow up and force the OOM to kick in?

So, the things _I've been talking about cloning_ specifically are parts of the execution graph, not "data" per se. Typically, they are quite small and model the steps that will execute :rofl:

The problematic part here WRT your query is "cloning these steps" is just another way of saying you'll be expending resources to repeat work already done as the join is re-evaluated over and over. If this were SQL - we're talking about something similar to a _subquery_, like when you run a `SELECT` per row within a larger query. It gets expensive.

Sorry I don't have better answers for you today. In the meantime, consider rewriting

```
intrinsic_value = join(tables: {oi: oi, strikes: strikes}, on: ["isin"], method: "inner")
```
as
```
// produces a "fresh" join each time...
intrinsic_value = () => join(tables: {oi: oi, strikes: strikes}, on: ["isin"], method: "inner")
```
then "call it" like a function
```
callsValueBelowPrice = (price) => {
  sum = intrinsic_value() // extra parens... 
    |> filter(fn: (r) => r.type == "call")
    |> filter(fn: (r) => float(v: r.price) == price)
    |> filter(fn: (r) => float(v: r.strike) < price)
    |> keep(columns: ["cash_value"])
    |> sum(column: "cash_value")
    |> findRecord(fn: (key) => true, idx: 0)

  return sum.cash_value
}
```

This should sidestep the panic and will probably perform similarly as the "fix" will, alas.

**Comment by Byter09** [comment]
Thank you for explaining this to me. I really appreciate it. I'll play around with that approach :D

**Comment by Byter09** [comment]
So, my current solution is to calculate the intrinsic values, push them to a new field in the same measurement using a task, and then call the code that uses functions to get values above and below the price. Long story short, for some reason the task takes ~15s and the query takes roughly 10s on my server. So a cumulative 25s. I have no clue how this could blow up and take over 8 minutes previously, lol.

Not a bad approach. Am essentially pre-processing stuff. Also learned a bit about tasks :D

**Comment by onelson** [comment]
> I have no clue how this could blow up and take over 8 minutes previously, lol.

I too am not sure. I captured a trace of the repro script and we can see that the `map()` call does indeed eat most of the runtime as expected, but what I didn't expect is for there to be such large variance from call to call.

![image](https://user-images.githubusercontent.com/301388/165798839-1302b2b7-bb86-468d-916e-09b81b9728e4.png)


My uneducated hunch is that this may have to do with garbage collection, and that the resources required by each run of a sub-program (via `findRecord`) are large enough to cause things to get backed up during execution.

We can likely open a separate issue to investigate the perf aspect of this, however this sort of usage of `findRecord` and friends is not particularly something we encourage for _other reasons_ besides perf (or so I'm told).

**Comment by Byter09** [comment]
If you ever investigate this, I'd love to get a mention or so, so that I can follow the development. I don't want to hog the issue list with "why is this so slow" :D

On the other hand I'm curious if the in-development Rust core would be any faster 👀

Oh and, if it's not encouraged, how else am I supposed to calculate this? Or are tasks and preprocessing the idiomatic way? But even if I preprocess, I still have to use findRecord to generate the value for the field. I'm a total flux noob, and this seemed like a solution I would've used in any other language as well.
Alternatively, I could probably pull the data into a separate tool, do the maths and then push new data back, but then the question remains of what even is the point of such a powerful tool like Flux if I have to go and look for alternative ways of doing this. I'd prefer to do this all in Flux so it's all in one place :D

**Comment by onelson** [comment]
> Oh and, if it's not encouraged, how else am I supposed to calculate this?

Tbh I'm also a flux noob :sweat_smile: so someone else will likely have better answers. I'm still learning.

I think part of what makes this particular solution problematic is that flux as it is today really wants to _not block_. This is to say, I think most flux programs are modeled as graphs that want to flow data through each operation incrementally -- stream processing.

When you have to block to evaluate some part of the graph to completion, that's harder to achieve. If you need to do that N times in a map, things get worse.

In a more general purpose lang, you'd be building a lookup table (once) and reusing the calculations during your map. This is what the `array.to` proposal would allow for.

I think when I was conferring with the team, someone mentioned perhaps using a 2nd join could be an option, but I haven't looked deeper at that. Something where you produce a stream of the sums and join them with your strikes on price -- could be worth considering.

> I'd prefer to do this all in Flux so it's all in one place :D

Strong agree. I think we can get there, but there are some cases that are tricky right now.
